api/handlers.go

```diff
@@ -112,23 +140,105 @@ func (api *API) userHandlerPUT(w http.ResponseWriter, req *http.Request, ps http
 
 // userChangePasswordHandler changes a user's password, given the old one is known.
 func (api *API) userChangePasswordHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-	// TODO Implement
-	WriteJSON(w, struct{ msg string }{"Not implemented."})
+	ok, err := isSelf(req, ps)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	if !ok {
+		WriteError(w, errors.New("you cannot access other users' info"), http.StatusBadRequest)
+		return
+	}
+
+	var u *database.User
+	// Fetch the user by their id.
+	if id := ps.ByName("id"); id != "" {
+		u, err = api.staticDB.UserByID(req.Context(), id)
+		if err != nil {
+			// This is a Bad Request and not an Internal Server Error because
+			// the client has supplied an invalid user id.
+			WriteError(w, errors.AddContext(err, "failed to fetch user"), http.StatusBadRequest)
+			return
+		}
+	}
+	err = req.ParseMultipartForm(MaxMultipartMem)
+	if err != nil {
+		WriteError(w, errors.New("Failed to parse multipart parameters."), http.StatusBadRequest)
+		return
+	}
+	// Validate that the given old password is correct.
+	if pw := req.PostFormValue("oldPassword"); pw != "" {
```

Comment:
Looks like if `pw == ""` it skips the validation, so someone can just omit it and override a password without knowing it

Reply:
Ha! Brilliant! Exactly the opposite behaviour than what we want! Thanks for that catch. :)